### PR TITLE
TRUNK-6030: Validating input String in the Allergen field ~ Allergy UI

### PR DESCRIPTION
**Pull request by Muta-Jonathan**
## Description of what I changed

After my changes, now our user is not allowed to enter any numeric or symbolic value in the Other Allergen section
 

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6030

## Steps to Implement the changes
* Build mvn clean install Openmrs-core
* Copy your openmrs. war file
* Give it a version like openmrs-2.4.0.war
* Then mvn openmrs-sdk:run to run the server

Thank you

### CODE REVIEW

**Comment by Muta-Jonathan** [comment]
cc @dkayiwa @sherrif10 ........... @mozzy11 ...... @ibacher ......please take a look

**Comment by Muta-Jonathan** [comment]
https://user-images.githubusercontent.com/86114366/134081066-e96f5d13-b997-4f7e-84c1-f54ab3838eef.mp4

A small video to show its implications ☝️

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:183]: Remove the unwanted spaces in this method

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:188]: Do you need this line?

File: api/src/main/java/org/openmrs/validator/AllergyValidator.java
Comment (sherrif10) [inline api/src/main/java/org/openmrs/validator/AllergyValidator.java:93]: This looks nice and good progress, Replace this validator message with **Other Allergen must not contain a number or symbol** the one you have  globally called

Comment (Muta-Jonathan) [reply api/src/main/java/org/openmrs/validator/AllergyValidator.java:93]: @sherrif10  according to @mozzy11 ..I was told its better to internationalize ... which this message "error.allergyapi.allergen.nonCodedAllergen.validateInputString" represents in the message.properties file
Hope this was already resolve

Comment (Muta-Jonathan) [reply api/src/main/java/org/openmrs/validator/AllergyValidator.java:93]: error.allergyapi.allergen.nonCodedAllergen.validateInputString=Other Allergen must not contain a number or symbol

.........Has u see up that is in the message.properties

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (Muta-Jonathan) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:188]: Yes... @HerbertYiga ..........so that the **allergen.class** is called in the **Allergy.class**

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (Muta-Jonathan) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:183]: @HerbertYiga ......indeed i also wanted to remove them but......according to the styling of the other Tests...........I just wanted to follow the Order.............Is it worth it.......to remove them or......just leave as the other tests

**Review by Muta-Jonathan — COMMENTED** [review]
Is it ouk now @sherrif10 ... @HerbertYiga

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (tendomart) [inline api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:190]: I honestly think you don't need this space

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (tendomart) [inline api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:192]: And this .

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (tendomart) [inline api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:183]: You may also want to rename the method to reflect the method you're testing like

`public void validate_shouldRejectNumericAllergenValue() {}`

Comment (Muta-Jonathan) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java]: Done the needful
Thanks @tendomart any other changes

File: api/src/main/resources/messages.properties
Comment (ibacher) [inline api/src/main/resources/messages.properties:95]:
```suggestion
error.allergyapi.allergen.nonCodedAllergen.cannotBeNumeric=Non-coded Allergen must not be only a number
```

File: api/src/main/java/org/openmrs/validator/AllergyValidator.java
Comment (ibacher) [inline api/src/main/java/org/openmrs/validator/AllergyValidator.java:94]: This is far too restrictive. For example, this forbids me from saying that I am allergic to "peanut butter", "Sputnik V COVID-19 Vaccine", "pâte d’arachide", never mind "Спутник V". All of those seem like valid allergens to me.

I'm not convinced that this ticket is actually a good idea, but if we're going to enforce something like the rule "an allergen cannot be solely a number" than the implementation should just limit us to that:

```java
if (NumberUtils.isParsable(allergen.getNonCodedAllergen()) {
	errors.rejectValue("allergen", "error.allergyapi.allergen.nonCodedAllergen.cannotBeNumeric");
}
```

**Review by ibacher — CHANGES_REQUESTED** [review]
@Muta-Jonathan Thanks for your work on this. I'm not convinced that this ticket is a great idea (see my comments below for why), but I've got some changes that should be made before we merge this, largely to be sure we aren't being too aggressive and that the error message key is properly restricted to only this error message.

File: api/src/main/java/org/openmrs/validator/AllergyValidator.java
Comment (Muta-Jonathan) [reply api/src/main/java/org/openmrs/validator/AllergyValidator.java:94]: Thanks, @ibacher ....in this aspect it really makes sense

**Comment by Muta-Jonathan** [comment]
cc @ibacher ...

When we talk about this validation of using 

**if (NumberUtils.isParsable(allergen.getNonCodedAllergen()) {
	errors.rejectValue("allergen", "error.allergyapi.allergen.nonCodedAllergen.cannotBeNumeric");
}**
its really bringing it out without even excluding other aspects like Covid-19.......Thanks a lot

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (tendomart) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java]: It is true , some other tests have spaces, but only were appropriate .Not everywhere

Comment (Muta-Jonathan) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java]: Thanks, @tendomart I think all issues a now solved........

**Comment by tendomart** [comment]
@ibacher how do we get github actions run the tests here ?

**Comment by gitcliff** [comment]
@tendomart you can run manually run the workflow since its a PR from a first time contributor......
workflows on pull requests to public repositories from some outside contributors will not run automatically, and might need to be approved first. By default, all first-time contributors require approval to run workflows

**Comment by Muta-Jonathan** [comment]
cc @ibacher

File: api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java
Comment (ibacher) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java:187]: I'm with @HerbertYiga on this. I don't really see what calling `allergy.getAllergen()` does here... all it does is return the `allergen` you just set by using `allergy.setAllergen()` and that result is ignored, so this is effectively dead code.

Otherwise, this looks fine. Thanks @Muta-Jonathan.

Comment (Muta-Jonathan) [reply api/src/test/java/org/openmrs/validator/AllergyValidatorTest.java]: Thanks, @ibacher ........have removed `allergy.getAllergen();`.........
@HerbertYiga have done the needful

**Review by tendomart — CHANGES_REQUESTED** [review]

**Review by tendomart — CHANGES_REQUESTED** [review]
Did you also test the message you introduced in the `messages.properties`  ?

and looks like you got some conflicts

**Comment by Muta-Jonathan** [comment]
Yes @tendomart and it works as expected

File: api/src/main/resources/messages.properties
Comment (Muta-Jonathan) [reply api/src/main/resources/messages.properties:95]: This helps to show the String assigned to` error.allergyapi.allergen.nonCodedAllergen.cannotBeNumeric` 
You may ask why not coded within.... you are right but due to internationalization...this was supposed to be done so
@tendomart hope I was helpful

**Comment by Muta-Jonathan** [comment]
Due to the wrong way to solve these conflicts @tendomart ...let me close and open a brand-new PR

File: api/src/main/resources/messages.properties
Comment (Muta-Jonathan) [reply api/src/main/resources/messages.properties:96]: > Did you also test the message you introduced in the `messages.properties` ?
> 
> and looks like you got some conflicts

**Pull request closed without merging**
